File: leetcode-75/18-dp-multidimensional/63-unique-paths/solution.py
"""
LeetCode 62. Unique Paths
https://leetcode.com/problems/unique-paths/

Approach: grid DP where dp[i][j] = dp[i-1][j] + dp[i][j-1], rolled down
to a single 1D row updated in place since each row only needs the row
above it and the value just computed to its left.

Time:  O(m * n) - every cell computed once
Space: O(n) - one rolling row instead of a full 2D table
"""


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        row = [1] * n

        for _ in range(1, m):
            for j in range(1, n):
                row[j] += row[j - 1]

        return row[-1]


def assert_equal(actual, expected, label):
    if actual != expected:
        print(f"FAIL [{label}]: got {actual}, expected {expected}")
    else:
        print(f"PASS [{label}]")


if __name__ == "__main__":
    sol = Solution()
    print(sol.uniquePaths(3, 7))  # 28
    print(sol.uniquePaths(3, 2))  # 3

    # Edge cases
    assert_equal(sol.uniquePaths(1, 1), 1, "m=1, n=1: start equals end")
    assert_equal(
        sol.uniquePaths(1, 5), 1, "m=1: single row, only path is straight right"
    )
    assert_equal(
        sol.uniquePaths(5, 1), 1, "n=1: single column, only path is straight down"
    )

File: leetcode-75/18-dp-multidimensional/63-unique-paths/solution.ts
/**
 * LeetCode 62. Unique Paths
 * https://leetcode.com/problems/unique-paths/
 *
 * Approach: grid DP where dp[i][j] = dp[i-1][j] + dp[i][j-1], rolled down
 * to a single 1D row updated in place since each row only needs the row
 * above it and the value just computed to its left.
 *
 * Time:  O(m * n) - every cell computed once
 * Space: O(n) - one rolling row instead of a full 2D table
 */
function uniquePaths(m: number, n: number): number {
  const row: number[] = new Array(n).fill(1);

  for (let i = 1; i < m; i++) {
    for (let j = 1; j < n; j++) {
      row[j] += row[j - 1];
    }
  }

  return row[n - 1];
}
function assertEqual(actual: unknown, expected: unknown, label: string): void {
  const a = JSON.stringify(actual);
  const e = JSON.stringify(expected);
  if (a !== e) {
    console.error(`FAIL [${label}]: got ${a}, expected ${e}`);
  } else {
    console.log(`PASS [${label}]`);
  }
}

if (require.main === module) {
  // Example usage:
  console.log(uniquePaths(3, 7)); // 28
  console.log(uniquePaths(3, 2)); // 3

  // Edge cases
  assertEqual(uniquePaths(1, 1), 1, 'm=1, n=1: start equals end');
  assertEqual(uniquePaths(1, 5), 1, 'm=1: single row, only path is straight right');
  assertEqual(uniquePaths(5, 1), 1, 'n=1: single column, only path is straight down');
}

export { uniquePaths };
